### PR TITLE
feat:  Allow JSON to be removed based on PAGS groups

### PR DESCRIPTION
**Pull request by cbeach47**
Based on the PAGS groups sent over via the OIDC token, this enhancement will see if the user is a member of of a set of specific groups.  If the user is a member, the specified JSON paths are removed from the JSON schema.

There is an inverse option as well - if the user is **not** a member of any of the specified groups, the JSON paths are removed from the JSON schema.

### CODE REVIEW

File: fbms-core/src/main/java/org/apereo/portal/fbms/data/filter/JsonSchemaExtensionFilter.java [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * If the user matches a set of PAGS group names (or the inverse),
+ * remove the specified JSON path(s)
+ */
+public class JsonSchemaExtensionFilter extends AbstractExtensionFilter<FbmsEntity> {
```
Comment (drewwills) [inline fbms-core/src/main/java/org/apereo/portal/fbms/data/filter/JsonSchemaExtensionFilter.java]: I think we need a better name -- many, many `ExtensionFilter` implementations will work with JSON Schema.

File: fbms-core/src/main/java/org/apereo/portal/fbms/data/filter/JsonSchemaExtensionFilter.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+    private static final Set<String> RELEVANT_HTTP_METHODS = new HashSet<>(Arrays.asList("GET"));
```
Comment (drewwills) [inline fbms-core/src/main/java/org/apereo/portal/fbms/data/filter/JsonSchemaExtensionFilter.java]: Try to use a constant for "GET" (Spring provides one).

File: fbms-core/src/main/java/org/apereo/portal/fbms/data/filter/JsonSchemaExtensionFilter.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        this.jsonToRemove = jsonToRemove;
+        this.inverseFlag = inverseFlag;
+    }
+
+    @Override
+    public boolean appliesTo(ExtensionFilterChainMetadata metadata, FbmsEntity entity, HttpServletRequest request) {
+        // Check if this form should use this filter
+        // targetForm must be supplied since we need a reference to the json structure
+        boolean appliesToForm = (this.targetForm != null) &&
+                !this.targetForm.isEmpty() &&
```
Comment (drewwills) [inline fbms-core/src/main/java/org/apereo/portal/fbms/data/filter/JsonSchemaExtensionFilter.java]: `StringUtils.isNotBlank(targetForm)`

File: fbms-core/src/main/java/org/apereo/portal/fbms/data/filter/JsonSchemaExtensionFilter.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+            if (associatedGroups.contains(tGroup)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private void removeJson(Map<String, List<String>> jsonPaths) {
+        // We need a reference to the form...
+        final FormEntity form = formRepository.findMostRecentByFname(this.targetForm);
```
Comment (drewwills) [inline fbms-core/src/main/java/org/apereo/portal/fbms/data/filter/JsonSchemaExtensionFilter.java]: The issue with the JSON removed for all users stems from making changes to the original `FormEntity` obtained from `findMostRecentByFname`.  You need to make a copy and alter the copy.

File: fbms-core/src/main/java/org/apereo/portal/fbms/data/filter/JsonSchemaExtensionFilter.java (hunk 1 above)
Comment (cbeach47) [reply fbms-core/src/main/java/org/apereo/portal/fbms/data/filter/JsonSchemaExtensionFilter.java]: While  a bit long, I changed it to a more specific name.

File: fbms-core/src/main/java/org/apereo/portal/fbms/data/filter/JsonSchemaExtensionFilter.java (hunk 2 above)
Comment (cbeach47) [reply fbms-core/src/main/java/org/apereo/portal/fbms/data/filter/JsonSchemaExtensionFilter.java]: :+1:

File: fbms-core/src/main/java/org/apereo/portal/fbms/data/filter/JsonSchemaExtensionFilter.java (hunk 3 above)
Comment (cbeach47) [reply fbms-core/src/main/java/org/apereo/portal/fbms/data/filter/JsonSchemaExtensionFilter.java]: :+1:

File: fbms-core/src/main/java/org/apereo/portal/fbms/data/filter/JsonSchemaExtensionFilter.java (hunk 4 above)
Comment (cbeach47) [reply fbms-core/src/main/java/org/apereo/portal/fbms/data/filter/JsonSchemaExtensionFilter.java]: Ah thanks - I switched to using the `formEntity.getSchema()` copy which isolated the JSON removal to the request.

File: fbms-core/src/main/java/org/apereo/portal/fbms/data/filter/JsonRemovalByPagsGroupsExtensionFilter.java [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * If the user matches a set of PAGS group names (or the inverse),
+ * remove the specified JSON path(s)
+ */
+public class JsonRemovalByPagsGroupsExtensionFilter extends AbstractExtensionFilter<FormEntity> {
```
Comment (drewwills) [inline fbms-core/src/main/java/org/apereo/portal/fbms/data/filter/JsonRemovalByPagsGroupsExtensionFilter.java]: How about `RoleBasedSchemaTrimmingExtensionFilter`?

PAGS is widely used in the portal for groups; but this `Filter` works with any sort of group, not just PAGS.

Comment (cbeach47) [reply fbms-core/src/main/java/org/apereo/portal/fbms/data/filter/JsonRemovalByPagsGroupsExtensionFilter.java]: Works for me.

**Pull request merged**